# **Framingham-Heart-Study**
## Trong dữ liệu này ta sẽ tiến hành huấn luyện một mô hình chuẩn đoán bệnh nhên có mắc bệnh tim mạch hay không

# **Một số đặc trưng cần chú ý**
- **sex** : Trong cột dữ liệu là male với 0 là nam, 1 là nữ.
- **age** : Tuổi của bệnh nhân
- **education** : Giáo dục với **với 1 là đang học trung học**, **2 là đã tốt nghiệp trung học**, **3 là Sinh viên**, **4 đã tốt nghiệp**
- **currentSmoker**: Tình trạng hút thuốc lá **(0 là ko hút, 1 là có hút)**
- **cigsPerDay**: Số lần hút thuốc trong ngày
- **BPmeds**: Có sử dụng thuốc giảm huyết áp trong quá trình kiểm tra hay không ? **(0 là không có, 1 là có)**
- **prevalentStroke**: Tỉ lệ mắc bệnh đột quỵ **(0 = Không là không mắc bệnh)**
- **prevalentHyp**: Tăng huyết áp hiện có **(0 là không tăng huyết áp, 1 là có tăng huyết áp)**
- **diabetes**: Có mắc bệnh tiểu đường hay không ? **(0 là không mắc, 1 là có mắc)**
- **totChol**: Tổng số cholesterol (mg/dL)
- **sysBP**: Huyết áp tâm thu (mmHg)
- **diaBP**: Huyết áp tâm trương (mmHg)
- **BMI**: Chỉ số khối cơ thể, cân nặng (kg)/chiều cao (m)^2
- **heartRate**: Nhịp tim (số nhịp/phút)
- **Glucose**: Nồng độ đường huyết (mg/dL)
- Và cuối cùng là biến phản hồi:
  + **TenYearCHD**: Nguy cơ mắc bệnh tim mạch vành (CHD) trong 10 năm. (0 là không mắc, 1 là có mắc bệnh)

In [1]:
import pandas as pd
data = pd.read_csv("./data/framingham.csv")
data.head()

,male,age,education,currentSmoker,cigsPerDay,BPMeds,prevalentStroke,prevalentHyp,diabetes,totChol,sysBP,diaBP,BMI,heartRate,glucose,TenYearCHD
0,1,39,4.0,0,0.0,0.0,0,0,0,195.0,106.0,70.0,26.97,80.0,77.0,0
1,0,46,2.0,0,0.0,0.0,0,0,0,250.0,121.0,81.0,28.73,95.0,76.0,0
2,1,48,1.0,1,20.0,0.0,0,0,0,245.0,127.5,80.0,25.34,75.0,70.0,0
3,0,61,3.0,1,30.0,0.0,0,1,0,225.0,150.0,95.0,28.58,65.0,103.0,1
4,0,46,3.0,1,23.0,0.0,0,0,0,285.0,130.0,84.0,23.10,85.0,85.0,0


In [2]:
#Check total columns and rows
data.shape

(4240, 16)

In [3]:
#Check name columns
data.columns

Index(['male', 'age', 'education', 'currentSmoker', 'cigsPerDay', 'BPMeds',
       'prevalentStroke', 'prevalentHyp', 'diabetes', 'totChol', 'sysBP',
       'diaBP', 'BMI', 'heartRate', 'glucose', 'TenYearCHD'],
      dtype='object')

In [4]:
#Check missing data and Dtype
data.info();

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4240 entries, 0 to 4239
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   male             4240 non-null   int64  
 1   age              4240 non-null   int64  
 2   education        4135 non-null   float64
 3   currentSmoker    4240 non-null   int64  
 4   cigsPerDay       4211 non-null   float64
 5   BPMeds           4187 non-null   float64
 6   prevalentStroke  4240 non-null   int64  
 7   prevalentHyp     4240 non-null   int64  
 8   diabetes         4240 non-null   int64  
 9   totChol          4190 non-null   float64
 10  sysBP            4240 non-null   float64
 11  diaBP            4240 non-null   float64
 12  BMI              4221 non-null   float64
 13  heartRate        4239 non-null   float64
 14  glucose          3852 non-null   float64
 15  TenYearCHD       4240 non-null   int64  
dtypes: float64(9), int64(7)
memory usage: 530.1 KB


In [5]:
#cigsPerDay is missing data
#education is missing data
#BPMeds is missing data
#totChol is missing data
#BMI is missing data
#heartRate is missing data
#glucose is missing data
data.isnull().sum()

male                 0
age                  0
education          105
currentSmoker        0
cigsPerDay          29
BPMeds              53
prevalentStroke      0
prevalentHyp         0
diabetes             0
totChol             50
sysBP                0
diaBP                0
BMI                 19
heartRate            1
glucose            388
TenYearCHD           0
dtype: int64

In [6]:
#Fix missing data
def impute_median(data):
    return data.fillna(data.median())


data.glucose = data['glucose'].transform(impute_median)
data.education = data['education'].transform(impute_median)
data.heartRate = data['heartRate'].transform(impute_median)
data.totChol = data['totChol'].transform(impute_median)
data.BPMeds = data['BPMeds'].transform(impute_median)

by_currentSmoker = data.groupby(['currentSmoker'])
data.cigsPerDay = by_currentSmoker['cigsPerDay'].transform(impute_median)

by_age = data.groupby(['male','age'])
data.BMI = by_age['BMI'].transform(impute_median)
data.isnull().sum()

male               0
age                0
education          0
currentSmoker      0
cigsPerDay         0
BPMeds             0
prevalentStroke    0
prevalentHyp       0
diabetes           0
totChol            0
sysBP              0
diaBP              0
BMI                0
heartRate          0
glucose            0
TenYearCHD         0
dtype: int64

In [7]:
#Explore descript data
data.describe()

,male,age,education,currentSmoker,cigsPerDay,BPMeds,prevalentStroke,prevalentHyp,diabetes,totChol,sysBP,diaBP,BMI,heartRate,glucose,TenYearCHD
count,4240.000000,4240.000000,4240.000000,4240.000000,4240.000000,4240.000000,4240.000000,4240.000000,4240.000000,4240.000000,4240.000000,4240.000000,4240.000000,4240.000000,4240.000000,4240.000000
mean,0.429245,49.580189,1.979953,0.494104,9.081132,0.029245,0.005896,0.310613,0.025708,236.667689,132.354599,82.897759,25.798756,75.878774,81.600943,0.151887
std,0.495027,8.572942,1.007087,0.500024,11.916119,0.168513,0.076569,0.462799,0.158280,44.328480,22.033300,11.910394,4.072774,12.023937,22.860340,0.358953
min,0.000000,32.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,107.000000,83.500000,48.000000,15.540000,44.000000,40.000000,0.000000
25%,0.000000,42.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,206.000000,117.000000,75.000000,23.077500,68.000000,72.000000,0.000000
50%,0.000000,49.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,234.000000,128.000000,82.000000,25.397500,75.000000,78.000000,0.000000
75%,1.000000,56.000000,3.000000,1.000000,20.000000,0.000000,0.000000,1.000000,0.000000,262.000000,144.000000,90.000000,28.040000,83.000000,85.000000,0.000000
max,1.000000,70.000000,4.000000,1.000000,70.000000,1.000000,1.000000,1.000000,1.000000,696.000000,295.000000,142.500000,56.800000,143.000000,394.000000,1.000000


# Tổng quan về dữ liệu 
- age : Max là 70, min là 32, mean là 49
- cigsPerDay : Max là 70 điếu thuốc, min là  0, mean là 9.08
- totChol : Max là 696, min là 44.32 , mean là 236
- sysBP : Max là 295, min là 83.5 , mean là 132
- diaBP : Max là 142, min là 48 , mean là 82.9
- BMI : Max là 56.8, min là 15.54 , mean là 25.7
- heartRate : Max là 143, min là 44 , mean là 75.8
- glucose : Max là 394, min là 40 , mean là 81.6

In [8]:
#Split 80% train and 20% test
from sklearn.model_selection import train_test_split
x = data.drop("TenYearCHD", axis=1)
y = data["TenYearCHD"]
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

In [9]:
#Scale data
from sklearn.preprocessing import StandardScaler
scale = StandardScaler()
x_train = scale.fit_transform(x_train)
x_test = scale.transform(x_test)

In [10]:
x_train

array([[-0.86045318,  0.39974402,  1.03053991, ..., -0.42165494,
         1.16336162,  0.57995512],
       [-0.86045318,  1.68620673,  2.02928031, ..., -1.40044805,
        -1.76502213, -0.38256091],
       [-0.86045318, -1.35452332,  1.03053991, ..., -0.89283574,
         0.74502108,  0.05494638],
       ...,
       [-0.86045318,  1.21840211,  0.03179952, ..., -0.17877823,
         0.57768487, -0.03255508],
       [ 1.16217828, -1.23757217,  0.03179952, ..., -0.32207549,
        -0.67733674, -0.86381893],
       [-0.86045318, -1.70537679,  0.03179952, ..., -1.25957956,
         1.16336162, -0.07630581]], shape=(3392, 15))

In [11]:
x_test

array([[-0.86045318, -0.06806061,  1.03053991, ..., -0.66210287,
        -1.09567728, -0.73256675],
       [ 1.16217828, -0.76976754, -0.96694088, ..., -0.48237411,
        -0.51000053,  0.40495221],
       [ 1.16217828, -0.53586523, -0.96694088, ...,  0.70772173,
        -0.59366863, -0.33881019],
       ...,
       [-0.86045318,  1.56925558, -0.96694088, ...,  0.51827791,
        -0.09165999,  0.22994929],
       [-0.86045318, -0.53586523, -0.96694088, ..., -0.38522343,
         1.33069783,  0.22994929],
       [ 1.16217828, -1.23757217, -0.96694088, ...,  0.66643269,
         1.16336162,  0.66745658]], shape=(848, 15))

In [12]:
#Training model and report
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report
model = RandomForestClassifier()
model = model.fit(x_train, y_train)
predict = model.predict(x_test)
report = classification_report(y_test, predict)
print(report)

              precision    recall  f1-score   support

           0       0.86      0.99      0.92       725
           1       0.62      0.07      0.12       123

    accuracy                           0.86       848
   macro avg       0.74      0.53      0.52       848
weighted avg       0.83      0.86      0.81       848



In [13]:
percent = model.predict_proba(x_test)
print(percent)

[[0.91 0.09]
 [0.92 0.08]
 [0.85 0.15]
 ...
 [0.78 0.22]
 [0.96 0.04]
 [0.97 0.03]]


In [14]:
check = data[data["TenYearCHD"] == 1]
check

,male,age,education,currentSmoker,cigsPerDay,BPMeds,prevalentStroke,prevalentHyp,diabetes,totChol,sysBP,diaBP,BMI,heartRate,glucose,TenYearCHD
3,0,61,3.0,1,30.0,0.0,0,1,0,225.0,150.0,95.0,28.58,65.0,103.0,1
6,0,63,1.0,0,0.0,0.0,0,0,0,205.0,138.0,71.0,33.11,60.0,85.0,1
15,0,38,2.0,1,20.0,0.0,0,1,0,221.0,140.0,90.0,21.35,95.0,70.0,1
17,0,46,2.0,1,20.0,0.0,0,0,0,291.0,112.0,78.0,23.38,80.0,89.0,1
25,1,47,4.0,1,20.0,0.0,0,0,0,294.0,102.0,68.0,24.18,62.0,66.0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4221,1,50,1.0,0,0.0,0.0,0,0,0,260.0,119.0,74.0,21.85,80.0,72.0,1
4223,1,56,4.0,0,0.0,1.0,0,1,0,287.0,149.0,98.0,21.68,90.0,75.0,1
4226,1,58,1.0,0,0.0,0.0,0,0,0,233.0,125.5,84.0,26.05,67.0,76.0,1
4232,1,68,1.0,0,0.0,0.0,0,1,0,176.0,168.0,97.0,23.14,60.0,79.0,1


In [15]:
patient_df = pd.DataFrame([{
    "male": 0,
    "age": 61,
    "education": 3.0,
    "currentSmoker": 1,
    "cigsPerDay": 30.0,
    "BPMeds": 0.0,
    "prevalentStroke": 0,
    "prevalentHyp": 1,
    "diabetes": 0,
    "totChol": 225.0,
    "sysBP": 150.0,
    "diaBP": 95.0,
    "BMI": 28.58,
    "heartRate": 65.0,
    "glucose": 103.0
}])

patient_scaled = scale.transform(patient_df)

prediction = model.predict(patient_scaled)
probability = model.predict_proba(patient_scaled)

print("Dự đoán:", prediction[0])
print("Xác suất mắc bệnh:", probability[0][1] * 100, "%")

Dự đoán: 1
Xác suất mắc bệnh: 77.0 %


In [16]:
patient_df = pd.DataFrame([{
    "male": 1,
    "age": 25,
    "education": 4.0,
    "currentSmoker": 0,
    "cigsPerDay": 0.0,
    "BPMeds": 0.0,
    "prevalentStroke": 0,
    "prevalentHyp": 0,
    "diabetes": 0,
    "totChol": 200.0,
    "sysBP": 150.0,
    "diaBP": 95.0,
    "BMI": 30.58,
    "heartRate": 65.0,
    "glucose": 103.0
}])

patient_scaled = scale.transform(patient_df)

prediction = model.predict(patient_scaled)
probability = model.predict_proba(patient_scaled)

print("Dự đoán:", prediction[0])
print("Xác suất mắc bệnh:", probability[0][1] * 100, "%")


Dự đoán: 0
Xác suất mắc bệnh: 14.000000000000002 %


In [19]:
#Xuat file pkl
import joblib

# Lưu mô hình Random Forest (ví dụ đây là mô hình tốt nhất của bạn)
joblib.dump(model, 'framing_ham_heart_disease_model.pkl')

# Lưu luôn cả bộ scale dữ liệu (nếu có dùng) để mai mốt scale dữ liệu mới nhập vào
joblib.dump(scale, 'scaler.pkl')

['scaler.pkl']